2.train-VAE/betavae.py

```diff
@@ -0,0 +1,115 @@
+        latent_dim, 
+        beta):
+        super(BetaVAE, self).__init__()
+        self.encoder = nn.Linear(input_dim, latent_dim * 2)
+        self.decoder = nn.Sequential(
+            nn.Linear(latent_dim, input_dim),
+            nn.Sigmoid()
+        )
+        self.latent_dim = latent_dim
+        self.beta = beta
+
+    def reparameterize(self, mu, log_var):
+        """
+        Reparameterize from N(mu, var) to N(0,1)
+        """
+        std = torch.exp(0.5 * log_var)
+        eps = torch.randn_like(std)
+        return mu + eps * std
+
+    def forward(self, x):
+        """
+        Forward pass through the VAE.
+
+        Args:
+           
+            x (Tensor): Input data.
+
+        Returns:
+            Reconstructed data, mean, and log variance of the latent space.
+        """
+        h = self.encoder(x)
+        mu, log_var = h.chunk(2, dim=-1)
+        z = self.reparameterize(mu, log_var)
+        return self.decoder(z), mu, log_var
+
+    def loss_function(self, recon_x, x, mu, log_var):
+        """
+        Compute the VAE loss function.
+
+        Args:
+            recon_x (Tensor): Reconstructed data.
+            x (Tensor): Original input data.
+            mu (Tensor): Mean of the latent Gaussian.
+            log_var (Tensor): Log variance of the latent Gaussian.
+
+        Returns:
+            Computed loss.
+        """
+        BCE = F.binary_cross_entropy(recon_x, x, reduction='sum')
+        KLD = -0.5 * torch.sum(1 + log_var - mu.pow(2) - log_var.exp())
+        return BCE + self.beta * KLD
+    
+def train_vae(model, train_loader, optimizer, epochs=5):
+    """
+    Train the VAE model.
+
+    Args:
+        model (VAE): VAE model to be trained.
+        train_loader (DataLoader): DataLoader for the training data.
+        optimizer (Optimizer): Optimizer for the model.
+        epochs (int, optional): Number of training epochs. Defaults to 5.
+    """
+    model.train()
+    for epoch in range(epochs):
+        train_loss = 0
+        for batch in train_loader:
+            data = batch[0]
+            optimizer.zero_grad()
+            recon_batch, mu, log_var = model(data)
+            loss = model.loss_function(recon_batch, data, mu, log_var)
+            loss.backward()
+            train_loss += loss.item()
+            optimizer.step()
+        print(f'Epoch {epoch}, Loss: {train_loss / len(train_loader.dataset)}')
+
+def evaluate_vae(model, test_loader):
+    """
+    Evaluate the VAE model.
+
+    Args:
+        model (VAE): VAE model to be evaluated.
+        test_loader (DataLoader): DataLoader for the test data.
+
+    Returns:
+        Average loss over the test dataset.
+    """
+    model.eval()
+    test_loss = 0
+    with torch.no_grad():
+        for batch in test_loader:
+            data = batch[0]
+            recon, mu, log_var = model(data)
+            test_loss += model.loss_function(recon, data, mu, log_var).item()
+    return test_loss / len(test_loader.dataset)
```

Comment:
this nasty symbol means that you need to add a new line character to the end of your file :) (black/autoformatters will completely solve this)